### ISSUE TITLE
Suggestions for improving documentation (JOSS review)

### ISSUE THREAD

**Issue by nluetts** [kellino/DiscreteEntropy.jl#13]
Here are some suggestions on how to improve the docs, again with reference to the JOSS review https://github.com/openjournals/joss-reviews/issues/7334,

This is not yet complete, but as far as I got today.

I added "[X]" to some of these points with X = A, B, C, ... so you can easily
reference them if you want to comment. The headings just follow the structure of your documentation.

# Overview

## Summary

[A] After "DiscreteEntropy implements a large collection of entropy
estimators.", add a list of all implemented estimators, so the user can get a
quick, central overview (otherwise one has to scroll through the "Estimators"
section where there is no table of contents that lists all estimators). Also,
list the measures that one can calculate with your package (cross entropy, KL
divergence, ...,  mutual information, conditional entropy, etc.) and link to
their specific API documentation.

In general, the documentation is very good in terms of how comprehensive
it is, but since most of it is API documentation, it is lacking guidance
beyond the "Overview" page.

## Installing Discrete Entropy

Add spaces before "higher are supported. ..."
to make the markdown list render correctly:

```md
1. If you have not done so already, install [Julia](https://julialang.org/downloads/). Julia 1.8 and
   higher are supported. Nightly is not (yet) supported.
```

[B] Mention the current limit to Julia "<=1.10".

## Basic usage

[C] Add a new subsection (##) "Contributing" and link to `CONTRIBUTING.md`.
This would be a good bonus:
If a contributor would want to add their own estimator, how would that work?
(What is the interface, or what functions would need to be implemented?)

# Data

## `CountData`

[D] Why is the total number of samples `CountData.N` represented with a
floating point number? Is it not always an integer? Is the same true for the
multiplicity matrix?

Below the matrix, change "We have 1 bin with ~~a~~ 4 elements" with
"We have 1 bin with 4 elements". This one "a" can make it really
confusing :-)

## `from_data`

"A 0 value in the ~~histgram~~ histogram ..."

# Estimate_H

## `estimate_h`

"Return the estimate**d** ~~in nats of~~ Shannon entropy of `data` **in nats**
using `estimator`."

"Note

While most calls to estimate_h take a CountData struct, this is not true for
every estimator, especially those that work directly over samples, or need the
original structure of the histogram.

For a complete list of methods of the function, try

```
julia> methods(estimate_h)
```
"

[E] The call to `methods` returns a long list of method calls that take a CountData
object and only a single call that takes a SampleVecor:

```
...
[25] estimate_h(data::SampleVector, estimator::Type{T}, ::Type{Bootstrap}; seed, reps, concentration) where T<:AbstractEstimator
...
```

However, in this case every estimator seems to be allowed
(`T<:AbstractEstimator`), so it does not get clear at this point
which estimators require which kind of data.

[F] Put

"This function is a wrapper indended to make using the libary easier. For finer
control over some of the estimators, it is advisable to call them directly,
rather than through this function."

right under

"Return the estimate in nats of Shannon entropy of data using estimator."

(or the revised version).

# Estimators

[G] I would advise putting the API reference for the abstract base types
at the end, since they are mainly of interest for contributors.
On the other hand, you should put the types for the concrete estimators here
(and in the overview as a list). Since you advertised `estimate_h` which needs
them as inputs earlier in the docs, you should provide these types now, since
one cannot always guess from the function names how the corresponding types are
called exactly (e.g. LaPlace --- is this  really how it should be written, by
the way?).

## `grassberger`

The link in

"Equation 13 from Finite sample corrections to entropy and dimension estimate"

yields a 404.

## `schurmann`

"~~This~~ There is no one ideal ..."

## `schurmann_generalized`

The link underneath the function signature should go to the end and
have "External Links" above it.

[H] "Computes the generalised Schurmann entropy estimation, given a countvector
data and a vector of xi values."

This says "vector of xi values" but the function signature above it says
`xis::Distribution`.

## `bub`

It would be good to explain the optional keyword arguments.

The external link goes to a blank page saying:

"Your session has timed out. Please go back to the article page and click the PDF link again."

(The URL contains a token that is invalid.)

## `chao_wang_jost`

The external link does not open for me.

## `unseen`

The docs do not render correctly for this one, seems a "`" is missing.

The external link requires an google account and access permission.

In general it would be good to give dois of the cited papers, 
so people can find them even when the links are not working.

## `minimax`

The math is not rendered correctly.

## `nsb`

The function signature does not include type information.

## `ansb`

Should say what keyword argument `undersampled` is (N/K?).

## `pym`

[I] This function does not conform to the standard API.
What are the arguments `mm` and `icts`? Why does it not
take CountData?

**Comment by nluetts** [comment]
Here is the rest of my suggestions.

@kellino I am pinging you here, in case you did not see this issue.


# Mixed Estimators

## `pert`

I would not use verbatim for explaining the parameter a,b,c here, because
usually it is only used for code examples.

[J] Explain the arguments. What is swapped with `estimator`, a or b or c?
What about `e1` and `e2`, a and c? Why can one not choose a, b and c
freely?

## `jackknife`

What is `corrected`?

## `bayesian_bootstrap`

Put keyword arguments `seed` and `concentration` in description in backticks.

The link is behind a paywall. Is there no paper that describes how
Bayesian bootstrapping works?

# Utility Functions

## `logx`

[K] I do not understand why returning 0 instead of -Inf (like the 
standard Julia function) is a good idea. You use this function in your
code, so perhaps you can explain why this is fine in the context of
entropy estimation?

## `bins`

Use verbatim for `x.multiplicities`

## `from_csv`

~~*CSV.File()~~ → `CSV.File()`

You can pass all further keyword arguments on to `CSV.File()` by changing the
function signature to

```julia
from_csv(file::String, field, ::Type{T}; remove_zeros=false, header=nothing, kw...)
```

and swapping `for row in CSV.File(file; header=header)` in the source code
with `for row in CSV.File(file; header=header, kw...)`

Then, users will have more options before falling back to using CSV.jl
directly.

# Divergence and Distance

## `cross_entropy`

[L] In the function signature, should it really be `{T<:MaximumLikelihood}`?
Looking at the source code, this should be a typo.

## `kl_divergence`

"~~Both distributions needs to be defined over the same space, so length(p) == length(q).~~"

Use the same statement as in the description of `cross_entropy`: "P and Q must be the same length."

"If the distributions are not over the same space **(their cross entropy is < 0)**, ~~then it~~the function return**s** Inf."

(cross entropy < 0 according to source code, I find this easier to understand)

## `jensen_shannon_divergence`

[N] The description says "P and Q, as represented by their histograms."
Why does the function take `AbstractVector`s and not `CountVector`s?

## `jensen_shannon_distance`

[O] Arguments are called `P` and `Q` like for `cross_entropy` and
`kl_divergence`, where they are of type `CountData`, but here they
are `AbstractVector`s. Should they be `CountVector`s, like for
`jensen_shannon_divergence`?

## `uncertainty_coefficient`

[P] What is `joint` in the equations in the description? *I(X;Y)*?

# Mutual Information and Conditional Information

## `mutual_information`

Description has some typo, does not render correctly:

"using Optim: estimatemaxstep mutualinformation(joint::Matrix{I}, estimator::Type{T}) where {T<:AbstractEstimator, I<:Real}"

(looks like a surplus or misplaced `using Optim: estimate_maxstep` in the docstring)

[Q] What is `XY` and where would it come from? I think this one
would benefit from an example.

## `conditional_entropy`

Explain the keyword arguments.

"and estimator ~~*estimator*~~`estimator`" (verbatim)

**Comment by kellino** [comment]
Hi @nluetts, this is an extraordinary effort on your part. Thank you very much indeed! I have pushed (to paper and main) and hope that I've covered every point that you've raised directly in the docs. I'm going to close this issue now, but please reopen if you think I've missed something or not done something very well.
